File: cpp/leetcode/037.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        dfs(board, 0, 0);
    }
    
    bool check(vector<vector<char>>& board, char ch, int row, int col){
        for(int i = 0; i < 9; i++){
            if(board[row][i] == ch || board[i][col] == ch) return false;
        }
        for(int i = row/3*3; i < row/3*3 + 3; i++){
            for(int j = col/3*3; j < col/3*3+3;j++){
                if(board[i][j] == ch) return false;
            }
        }
        return true;
    }
    
    bool dfs(vector<vector<char>>& board, int row, int col){
        int nrow = -1, ncol = -1;
        if(row == 9) return true;
        if(col == 8){
            nrow = row + 1;
            ncol = 0;
        } else {
            nrow = row;
            ncol = col + 1;
        }
        if(board[row][col] != '.'){
            return dfs(board, nrow, ncol);
        }
        for(int ch = 0; ch < 9; ch++){
            if(check(board, ch + '1', row, col)){
                board[row][col] = ch + '1';
                if(dfs(board, nrow, ncol))  return true;
                board[row][col] = '.';
            }
        }
        return false;
    }
};

File: python/leetcode/037.py
class Solution(object):
    def solveSudoku(self, board):
        self.dfs(board,  0, 0)

    def check(self, board, char, row, col):
        for i in xrange(9):
            if board[row][i] == char or board[i][col] == char:
                return False
        for i in xrange(row/3*3, row/3*3+3):
            for j in xrange(col/3*3, col/3*3+3):
                if board[i][j] == char:
                    return False
        return True

    def dfs(self, board, row, col):
        if row == 9: return True
        if col == 8:
            nextRow = row + 1
            nextCol = 0
        else:
            nextRow = row
            nextCol = col+1
        if board[row][col] != '.': #occupied
            return self.dfs(board, nextRow, nextCol)
        for c in range(1, 10): # potential number
            if self.check(board, str(c), row, col):  
                board[row][col] = str(c)
                if self.dfs(board, nextRow, nextCol):  return True
                board[row][col] = '.'
        return False
